lib/Transforms/Scalar/DxilFixConstArrayInitializer.cpp

```diff
@@ -0,0 +1,184 @@
+    return false;
+
+  // Only handle cases when it's a scalar type or if it's an array of
+  // scalars.
+  Type *Ty = GV.getType()->getPointerElementType();
+  if (Ty->isArrayTy() && Ty->getArrayElementType()->isArrayTy())
+    return false;
+
+  // Record the earliest load and latest store in the entry block.
+  unsigned EarliestLoad = std::numeric_limits<unsigned>::max();
+  unsigned LatestStore = 0;
+
+  bool IsArray = Ty->isArrayTy();
+  SmallVector<Constant *, 16> InitValue;
+  SmallVector<unsigned, 16> LatestStores;
+
+  Type *ElementTy = nullptr;
+  if (IsArray) {
+    InitValue.resize(Ty->getArrayNumElements());
+    LatestStores.resize(Ty->getArrayNumElements());
+    ElementTy = Ty->getArrayElementType();
+  }
+  else {
+    InitValue.resize(1);
+    LatestStores.resize(1);
+    ElementTy = Ty;
+  }
+
+  SmallVector<StoreInst *, 8> StoresToRemove;
+
+  for (User *U : GV.users()) {
+    if (GEPOperator *GEP = dyn_cast<GEPOperator>(U)) {
+      bool AllConstIndices = GEP->hasAllConstantIndices();
+      unsigned NumIndices = GEP->getNumIndices();
+      if (IsArray && NumIndices != 2)
+        return false;
+      if (!IsArray && NumIndices != 1)
+        return false;
+
+      for (User *GEPUser : GEP->users()) {
+        if (StoreInst *Store = dyn_cast<StoreInst>(GEPUser)) {
+          if (Store->getParent() == EntryBlock) {
+            if (!isa<Constant>(Store->getValueOperand()))
+              return false;
+
+            if (!AllConstIndices)
+              return false;
+
+            unsigned StoreIndex = InstOrder.at(Store);
+            if (StoreIndex > EarliestLoad)
+              return false;
+
+            LatestStore = std::max(LatestStore, StoreIndex);
+
+            if (IsArray) {
+              uint64_t Index = cast<ConstantInt>(GEP->getOperand(2))->getLimitedValue();
+              if (LatestStores[Index] <= StoreIndex) {
+                InitValue[Index] = cast<Constant>(Store->getValueOperand());
+                LatestStores[Index] = StoreIndex;
+              }
+            }
+            else {
+              if (LatestStores[0] <= StoreIndex) {
+                InitValue[0] = cast<Constant>(Store->getValueOperand());
+                LatestStores[0] = StoreIndex;
+              }
+            }
+
+            StoresToRemove.push_back(Store);
+          }
+        }
+        else if (LoadInst *Load = dyn_cast<LoadInst>(GEPUser)) {
+          if (Load->getParent() == EntryBlock)
+            EarliestLoad = std::min(EarliestLoad, InstOrder.at(Load));
+
+          if (EarliestLoad < LatestStore)
+            return false;
+        }
+        // If we have some weird cases, like chained GEPS, or bitcasts, give up.
+        else {
+          return false;
+        }
+      }
+    }
+  }
+
+  Constant *Initializer = nullptr;
+
+  for (Constant *C : InitValue)
+    if (!C)
+      return false;
+
+  if (IsArray) {
+    Initializer = ConstantArray::get(cast<ArrayType>(Ty), InitValue);
```

Comment:
I don't think you're guaranteed that none of InitValue are nullptr at this point. You should have a test where your array is initialized in your shader entry point rather than using the initializer syntax, and you could test what happens if you don't initialize all elements. Also it would be good to test the load-before-store logic.